In [7]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [8]:
pd.set_option("display.max_columns", None)

In [10]:
orders = pd.read_csv(
    "../data/raw/olist_orders_dataset.csv"
)

In [6]:
orders.head()

,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date
0,e481f51cbdc54678b7cc49136f2d6af7,9ef432eb6251297304e76186b10a928d,delivered,2017-10-02 10:56:33,2017-10-02 11:07:15,2017-10-04 19:55:00,2017-10-10 21:25:13,2017-10-18 00:00:00
1,53cdb2fc8bc7dce0b6741e2150273451,b0830fb4747a6c6d20dea0b8c802d7ef,delivered,2018-07-24 20:41:37,2018-07-26 03:24:27,2018-07-26 14:31:00,2018-08-07 15:27:45,2018-08-13 00:00:00
2,47770eb9100c2d0c44946d9cf07ec65d,41ce2a54c0b03bf3443c3d931a367089,delivered,2018-08-08 08:38:49,2018-08-08 08:55:23,2018-08-08 13:50:00,2018-08-17 18:06:29,2018-09-04 00:00:00
3,949d5b44dbf5de918fe9c16f97b45f8a,f88197465ea7920adcdbec7375364d82,delivered,2017-11-18 19:28:06,2017-11-18 19:45:59,2017-11-22 13:39:59,2017-12-02 00:28:42,2017-12-15 00:00:00
4,ad21c59c0840e6cb83a9ceb5573f8159,8ab97904e6daea8866dbdbc4fb7aad2c,delivered,2018-02-13 21:18:39,2018-02-13 22:20:29,2018-02-14 19:46:34,2018-02-16 18:17:02,2018-02-26 00:00:00


In [7]:
orders.shape

(99441, 8)

In [8]:
orders.columns

Index(['order_id', 'customer_id', 'order_status', 'order_purchase_timestamp',
       'order_approved_at', 'order_delivered_carrier_date',
       'order_delivered_customer_date', 'order_estimated_delivery_date'],
      dtype='str')

In [9]:
orders.info()

<class 'pandas.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 8 columns):
 #   Column                         Non-Null Count  Dtype
---  ------                         --------------  -----
 0   order_id                       99441 non-null  str  
 1   customer_id                    99441 non-null  str  
 2   order_status                   99441 non-null  str  
 3   order_purchase_timestamp       99441 non-null  str  
 4   order_approved_at              99281 non-null  str  
 5   order_delivered_carrier_date   97658 non-null  str  
 6   order_delivered_customer_date  96476 non-null  str  
 7   order_estimated_delivery_date  99441 non-null  str  
dtypes: str(8)
memory usage: 6.1 MB


In [10]:
orders.isnull().sum()

order_id                            0
customer_id                         0
order_status                        0
order_purchase_timestamp            0
order_approved_at                 160
order_delivered_carrier_date     1783
order_delivered_customer_date    2965
order_estimated_delivery_date       0
dtype: int64

In [11]:
orders.isnull().mean() * 100

order_id                         0.000000
customer_id                      0.000000
order_status                     0.000000
order_purchase_timestamp         0.000000
order_approved_at                0.160899
order_delivered_carrier_date     1.793023
order_delivered_customer_date    2.981668
order_estimated_delivery_date    0.000000
dtype: float64

In [13]:
orders.duplicated().sum()

np.int64(0)

In [14]:
orders["order_id"].duplicated().sum()

np.int64(0)

In [15]:
orders["order_id"].nunique()

99441

In [16]:
orders.shape[0]

99441

In [17]:
orders.columns

Index(['order_id', 'customer_id', 'order_status', 'order_purchase_timestamp',
       'order_approved_at', 'order_delivered_carrier_date',
       'order_delivered_customer_date', 'order_estimated_delivery_date'],
      dtype='str')

In [18]:
date_columns = [
    "order_purchase_timestamp",
    "order_approved_at",
    "order_delivered_carrier_date",
    "order_delivered_customer_date",
    "order_estimated_delivery_date"
]

for column in date_columns:
    orders[column] = pd.to_datetime(orders[column])

In [19]:
orders.info()

<class 'pandas.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 8 columns):
 #   Column                         Non-Null Count  Dtype         
---  ------                         --------------  -----         
 0   order_id                       99441 non-null  str           
 1   customer_id                    99441 non-null  str           
 2   order_status                   99441 non-null  str           
 3   order_purchase_timestamp       99441 non-null  datetime64[us]
 4   order_approved_at              99281 non-null  datetime64[us]
 5   order_delivered_carrier_date   97658 non-null  datetime64[us]
 6   order_delivered_customer_date  96476 non-null  datetime64[us]
 7   order_estimated_delivery_date  99441 non-null  datetime64[us]
dtypes: datetime64[us](5), str(3)
memory usage: 6.1 MB


In [20]:
(orders["order_approved_at"] < orders["order_purchase_timestamp"]).sum()

np.int64(0)

In [21]:
(orders["order_delivered_customer_date"] < orders["order_purchase_timestamp"]).sum()

np.int64(0)

In [22]:
orders[
    orders["order_approved_at"].isna()
]["order_status"].value_counts()

order_status
canceled     141
delivered     14
created        5
Name: count, dtype: int64

In [23]:
orders[
    orders["order_delivered_carrier_date"].isna()
]["order_status"].value_counts()

order_status
unavailable    609
canceled       550
invoiced       314
processing     301
created          5
approved         2
delivered        2
Name: count, dtype: int64

In [24]:
orders[
    orders["order_delivered_customer_date"].isna()
]["order_status"].value_counts()

order_status
shipped        1107
canceled        619
unavailable     609
invoiced        314
processing      301
delivered         8
created           5
approved          2
Name: count, dtype: int64

In [25]:
delivered_without_customer_date = orders[
    (orders["order_status"] == "delivered") &
    (orders["order_delivered_customer_date"].isna())
]

delivered_without_customer_date

,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date
3002,2d1e2d5bf4dc7227b3bfebb81328c15f,ec05a6d8558c6455f0cbbd8a420ad34f,delivered,2017-11-28 17:44:07,2017-11-28 17:56:40,2017-11-30 18:12:23,NaT,2017-12-18
20618,f5dd62b788049ad9fc0526e3ad11a097,5e89028e024b381dc84a13a3570decb4,delivered,2018-06-20 06:58:43,2018-06-20 07:19:05,2018-06-25 08:05:00,NaT,2018-07-16
43834,2ebdfc4f15f23b91474edf87475f108e,29f0540231702fda0cfdee0a310f11aa,delivered,2018-07-01 17:05:11,2018-07-01 17:15:12,2018-07-03 13:57:00,NaT,2018-07-30
79263,e69f75a717d64fc5ecdfae42b2e8e086,cfda40ca8dd0a5d486a9635b611b398a,delivered,2018-07-01 22:05:55,2018-07-01 22:15:14,2018-07-03 13:57:00,NaT,2018-07-30
82868,0d3268bad9b086af767785e3f0fc0133,4f1d63d35fb7c8999853b2699f5c7649,delivered,2018-07-01 21:14:02,2018-07-01 21:29:54,2018-07-03 09:28:00,NaT,2018-07-24
92643,2d858f451373b04fb5c984a1cc2defaf,e08caf668d499a6d643dafd7c5cc498a,delivered,2017-05-25 23:22:43,2017-05-25 23:30:16,NaT,NaT,2017-06-23
97647,ab7c89dc1bf4a1ead9d6ec1ec8968a84,dd1b84a7286eb4524d52af4256c0ba24,delivered,2018-06-08 12:09:39,2018-06-08 12:36:39,2018-06-12 14:10:00,NaT,2018-06-26
98038,20edc82cf5400ce95e1afacc25798b31,28c37425f1127d887d7337f284080a0f,delivered,2018-06-27 16:09:12,2018-06-27 16:29:30,2018-07-03 19:26:00,NaT,2018-07-19


In [26]:
delivered_without_customer_date[
    [
        "order_id",
        "customer_id",
        "order_status",
        "order_purchase_timestamp",
        "order_approved_at",
        "order_delivered_carrier_date",
        "order_delivered_customer_date",
        "order_estimated_delivery_date"
    ]
]

,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date
3002,2d1e2d5bf4dc7227b3bfebb81328c15f,ec05a6d8558c6455f0cbbd8a420ad34f,delivered,2017-11-28 17:44:07,2017-11-28 17:56:40,2017-11-30 18:12:23,NaT,2017-12-18
20618,f5dd62b788049ad9fc0526e3ad11a097,5e89028e024b381dc84a13a3570decb4,delivered,2018-06-20 06:58:43,2018-06-20 07:19:05,2018-06-25 08:05:00,NaT,2018-07-16
43834,2ebdfc4f15f23b91474edf87475f108e,29f0540231702fda0cfdee0a310f11aa,delivered,2018-07-01 17:05:11,2018-07-01 17:15:12,2018-07-03 13:57:00,NaT,2018-07-30
79263,e69f75a717d64fc5ecdfae42b2e8e086,cfda40ca8dd0a5d486a9635b611b398a,delivered,2018-07-01 22:05:55,2018-07-01 22:15:14,2018-07-03 13:57:00,NaT,2018-07-30
82868,0d3268bad9b086af767785e3f0fc0133,4f1d63d35fb7c8999853b2699f5c7649,delivered,2018-07-01 21:14:02,2018-07-01 21:29:54,2018-07-03 09:28:00,NaT,2018-07-24
92643,2d858f451373b04fb5c984a1cc2defaf,e08caf668d499a6d643dafd7c5cc498a,delivered,2017-05-25 23:22:43,2017-05-25 23:30:16,NaT,NaT,2017-06-23
97647,ab7c89dc1bf4a1ead9d6ec1ec8968a84,dd1b84a7286eb4524d52af4256c0ba24,delivered,2018-06-08 12:09:39,2018-06-08 12:36:39,2018-06-12 14:10:00,NaT,2018-06-26
98038,20edc82cf5400ce95e1afacc25798b31,28c37425f1127d887d7337f284080a0f,delivered,2018-06-27 16:09:12,2018-06-27 16:29:30,2018-07-03 19:26:00,NaT,2018-07-19


In [27]:
delivered_without_customer_date[
    [
        "order_purchase_timestamp",
        "order_approved_at",
        "order_delivered_carrier_date",
        "order_delivered_customer_date"
    ]
].isna()

,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date
3002,False,False,False,True
20618,False,False,False,True
43834,False,False,False,True
79263,False,False,False,True
82868,False,False,False,True
92643,False,False,True,True
97647,False,False,False,True
98038,False,False,False,True


,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date
3002,False,False,False,True
20618,False,False,False,True
43834,False,False,False,True
79263,False,False,False,True
82868,False,False,False,True
92643,False,False,True,True
97647,False,False,False,True
98038,False,False,False,True


In [29]:
delivered_without_customer_date[
    "order_delivered_carrier_date"
].notna().value_counts()

order_delivered_carrier_date
True     7
False    1
Name: count, dtype: int64

In [30]:
delivered_without_customer_date[
    "order_estimated_delivery_date"
].isna().sum()

np.int64(0)

In [37]:
delivered_without_carrier_date = orders[
    (orders["order_status"] == "delivered") &
    (orders["order_delivered_carrier_date"].isna())
]

In [32]:
delivered_without_carrier_date

,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date
73222,2aa91108853cecb43c84a5dc5b277475,afeb16c7f46396c0ed54acb45ccaaa40,delivered,2017-09-29 08:52:58,2017-09-29 09:07:16,NaT,2017-11-20 19:44:47,2017-11-14
92643,2d858f451373b04fb5c984a1cc2defaf,e08caf668d499a6d643dafd7c5cc498a,delivered,2017-05-25 23:22:43,2017-05-25 23:30:16,NaT,NaT,2017-06-23


In [33]:
delivered_without_carrier_date[
    [
        "order_id",
        "customer_id",
        "order_status",
        "order_purchase_timestamp",
        "order_approved_at",
        "order_delivered_carrier_date",
        "order_delivered_customer_date",
        "order_estimated_delivery_date"
    ]
]

,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date
73222,2aa91108853cecb43c84a5dc5b277475,afeb16c7f46396c0ed54acb45ccaaa40,delivered,2017-09-29 08:52:58,2017-09-29 09:07:16,NaT,2017-11-20 19:44:47,2017-11-14
92643,2d858f451373b04fb5c984a1cc2defaf,e08caf668d499a6d643dafd7c5cc498a,delivered,2017-05-25 23:22:43,2017-05-25 23:30:16,NaT,NaT,2017-06-23


In [34]:
delivered_without_carrier_date[
    "order_delivered_customer_date"
].notna().value_counts()

order_delivered_customer_date
True     1
False    1
Name: count, dtype: int64

In [35]:
delivered_without_customer_date[
    [
        "order_id",
        "order_purchase_timestamp",
        "order_approved_at",
        "order_delivered_carrier_date",
        "order_delivered_customer_date",
        "order_estimated_delivery_date"
    ]
].T

,3002,20618,43834,79263,82868,92643,97647,98038
order_id,2d1e2d5bf4dc7227b3bfebb81328c15f,f5dd62b788049ad9fc0526e3ad11a097,2ebdfc4f15f23b91474edf87475f108e,e69f75a717d64fc5ecdfae42b2e8e086,0d3268bad9b086af767785e3f0fc0133,2d858f451373b04fb5c984a1cc2defaf,ab7c89dc1bf4a1ead9d6ec1ec8968a84,20edc82cf5400ce95e1afacc25798b31
order_purchase_timestamp,2017-11-28 17:44:07,2018-06-20 06:58:43,2018-07-01 17:05:11,2018-07-01 22:05:55,2018-07-01 21:14:02,2017-05-25 23:22:43,2018-06-08 12:09:39,2018-06-27 16:09:12
order_approved_at,2017-11-28 17:56:40,2018-06-20 07:19:05,2018-07-01 17:15:12,2018-07-01 22:15:14,2018-07-01 21:29:54,2017-05-25 23:30:16,2018-06-08 12:36:39,2018-06-27 16:29:30
order_delivered_carrier_date,2017-11-30 18:12:23,2018-06-25 08:05:00,2018-07-03 13:57:00,2018-07-03 13:57:00,2018-07-03 09:28:00,NaT,2018-06-12 14:10:00,2018-07-03 19:26:00
order_delivered_customer_date,NaT,NaT,NaT,NaT,NaT,NaT,NaT,NaT
order_estimated_delivery_date,2017-12-18 00:00:00,2018-07-16 00:00:00,2018-07-30 00:00:00,2018-07-30 00:00:00,2018-07-24 00:00:00,2017-06-23 00:00:00,2018-06-26 00:00:00,2018-07-19 00:00:00


In [36]:
delivered_without_carrier_date[
    [
        "order_id",
        "order_purchase_timestamp",
        "order_approved_at",
        "order_delivered_carrier_date",
        "order_delivered_customer_date",
        "order_estimated_delivery_date"
    ]
].T

,73222,92643
order_id,2aa91108853cecb43c84a5dc5b277475,2d858f451373b04fb5c984a1cc2defaf
order_purchase_timestamp,2017-09-29 08:52:58,2017-05-25 23:22:43
order_approved_at,2017-09-29 09:07:16,2017-05-25 23:30:16
order_delivered_carrier_date,NaT,NaT
order_delivered_customer_date,2017-11-20 19:44:47,NaT
order_estimated_delivery_date,2017-11-14 00:00:00,2017-06-23 00:00:00


Auditoria da tabela orders: a tabela possui 99.441 registros e 8 atributos, apresentando order_id único para cada registro e ausência de duplicidades completas. Os identificadores de pedido e cliente não apresentam valores ausentes. As variáveis temporais inicialmente armazenadas como texto foram convertidas para datetime. Foram identificados valores ausentes em etapas intermediárias do ciclo do pedido, especialmente na data de entrega ao cliente (2,98%), cuja ausência deverá ser interpretada considerando o status do pedido antes de qualquer tratamento. A primeira validação temporal não identificou registros em que a aprovação ocorreu antes da compra.

## Exceções de consistência — pedidos entregues

Durante a auditoria da tabela `orders`, foram identificados registros com
`order_status = "delivered"` que apresentam ausência em etapas do fluxo
logístico.

- 8 pedidos possuem status `delivered`, mas não possuem
  `order_delivered_customer_date`.
- 2 pedidos possuem status `delivered`, mas não possuem
  `order_delivered_carrier_date`.
- Existe sobreposição entre os grupos: o pedido `92643` pertence aos dois
  grupos, totalizando 9 pedidos distintos afetados.

Os registros foram mantidos sem imputação ou exclusão, pois não há
informação suficiente para reconstruir as datas ausentes de forma confiável.

Essas ocorrências serão consideradas exceções de completude/consistência
durante as análises posteriores.

In [11]:
order_items = pd.read_csv(
    "../data/raw/olist_order_items_dataset.csv"
)

In [39]:
order_items.shape

(112650, 7)

In [40]:
order_items.columns

Index(['order_id', 'order_item_id', 'product_id', 'seller_id',
       'shipping_limit_date', 'price', 'freight_value'],
      dtype='str')

In [41]:
order_items.head()

,order_id,order_item_id,product_id,seller_id,shipping_limit_date,price,freight_value
0,00010242fe8c5a6d1ba2dd792cb16214,1,4244733e06e7ecb4970a6e2683c13e61,48436dade18ac8b2bce089ec2a041202,2017-09-19 09:45:35,58.90,13.29
1,00018f77f2f0320c557190d7a144bdd3,1,e5f2d52b802189ee658865ca93d83a8f,dd7ddc04e1b6c2c614352b383efe2d36,2017-05-03 11:05:13,239.90,19.93
2,000229ec398224ef6ca0657da4fc703e,1,c777355d18b72b67abbeef9df44fd0fd,5b51032eddd242adc84c38acab88f23d,2018-01-18 14:48:30,199.00,17.87
3,00024acbcdf0a6daa1e931b038114c75,1,7634da152a4610f1595efa32f14722fc,9d7a1d34a5052409006425275ba1c2b4,2018-08-15 10:10:18,12.99,12.79
4,00042b26cf59d7ce69dfabb4e55b4fd9,1,ac6c3623068f30de03045865e4e10089,df560393f3a51e74553ab94004ba5c87,2017-02-13 13:57:51,199.90,18.14


In [42]:
order_items.info()

<class 'pandas.DataFrame'>
RangeIndex: 112650 entries, 0 to 112649
Data columns (total 7 columns):
 #   Column               Non-Null Count   Dtype  
---  ------               --------------   -----  
 0   order_id             112650 non-null  str    
 1   order_item_id        112650 non-null  int64  
 2   product_id           112650 non-null  str    
 3   seller_id            112650 non-null  str    
 4   shipping_limit_date  112650 non-null  str    
 5   price                112650 non-null  float64
 6   freight_value        112650 non-null  float64
dtypes: float64(2), int64(1), str(4)
memory usage: 6.0 MB


In [43]:
order_items.isnull().sum()

order_id               0
order_item_id          0
product_id             0
seller_id              0
shipping_limit_date    0
price                  0
freight_value          0
dtype: int64

In [44]:
order_items.isnull().mean() * 100

order_id               0.0
order_item_id          0.0
product_id             0.0
seller_id              0.0
shipping_limit_date    0.0
price                  0.0
freight_value          0.0
dtype: float64

In [62]:
order_items["shipping_limit_date"] = pd.to_datetime(
    order_items["shipping_limit_date"]
)

In [63]:
order_items.info()

<class 'pandas.DataFrame'>
RangeIndex: 112650 entries, 0 to 112649
Data columns (total 7 columns):
 #   Column               Non-Null Count   Dtype         
---  ------               --------------   -----         
 0   order_id             112650 non-null  str           
 1   order_item_id        112650 non-null  int64         
 2   product_id           112650 non-null  str           
 3   seller_id            112650 non-null  str           
 4   shipping_limit_date  112650 non-null  datetime64[us]
 5   price                112650 non-null  float64       
 6   freight_value        112650 non-null  float64       
dtypes: datetime64[us](1), float64(2), int64(1), str(3)
memory usage: 6.0 MB


In [45]:
order_items.duplicated().sum()

np.int64(0)

In [64]:
order_items["order_id"].duplicated().sum()

np.int64(13984)

In [47]:
order_items["order_id"].nunique()

98666

In [65]:
order_items[
    ["order_id", "order_item_id"]
].duplicated().sum()

np.int64(0)

In [48]:
order_items.shape[0]

112650

In [49]:
items_per_order = order_items.groupby("order_id").size()

items_per_order.describe()

count    98666.000000
mean         1.141731
std          0.538452
min          1.000000
25%          1.000000
50%          1.000000
75%          1.000000
max         21.000000
dtype: float64

In [50]:
order_items[
    ["price", "freight_value"]
].dtypes

price            float64
freight_value    float64
dtype: object

In [52]:
order_items[
    ["price", "freight_value"]
].describe()

,price,freight_value
count,112650.000000,112650.000000
mean,120.653739,19.990320
std,183.633928,15.806405
min,0.850000,0.000000
25%,39.900000,13.080000
50%,74.990000,16.260000
75%,134.900000,21.150000
max,6735.000000,409.680000


In [67]:
order_items["price"].quantile(
    [0.01, 0.05, 0.25, 0.50, 0.75, 0.95, 0.99]
)

0.01      9.99
0.05     17.00
0.25     39.90
0.50     74.99
0.75    134.90
0.95    349.90
0.99    890.00
Name: price, dtype: float64

In [66]:
order_items["price"].quantile(
    [0.01, 0.05, 0.25, 0.50, 0.75, 0.95, 0.99]
)

0.01      9.99
0.05     17.00
0.25     39.90
0.50     74.99
0.75    134.90
0.95    349.90
0.99    890.00
Name: price, dtype: float64

In [53]:
(order_items["freight_value"] < 0).sum()

np.int64(0)

In [54]:
(order_items["price"] == 0).sum()

np.int64(0)

In [55]:
(order_items["freight_value"] == 0).sum()

np.int64(383)

In [56]:
order_items["product_id"].nunique()

32951

In [57]:
order_items["seller_id"].nunique()

3095

In [68]:
product_frequency = (
    order_items["product_id"]
    .value_counts()
)

product_frequency.describe()

count    32951.000000
mean         3.418713
std         10.619709
min          1.000000
25%          1.000000
50%          1.000000
75%          3.000000
max        527.000000
Name: count, dtype: float64

In [59]:
order_items["shipping_limit_date"].dtype

<StringDtype(storage='python', na_value=nan)>

In [69]:
order_items["shipping_limit_date"] = pd.to_datetime(
    order_items["shipping_limit_date"]
)

In [70]:
order_items.info()

<class 'pandas.DataFrame'>
RangeIndex: 112650 entries, 0 to 112649
Data columns (total 7 columns):
 #   Column               Non-Null Count   Dtype         
---  ------               --------------   -----         
 0   order_id             112650 non-null  str           
 1   order_item_id        112650 non-null  int64         
 2   product_id           112650 non-null  str           
 3   seller_id            112650 non-null  str           
 4   shipping_limit_date  112650 non-null  datetime64[us]
 5   price                112650 non-null  float64       
 6   freight_value        112650 non-null  float64       
dtypes: datetime64[us](1), float64(2), int64(1), str(3)
memory usage: 6.0 MB


In [71]:
Q1_price = order_items["price"].quantile(0.25)

Q3_price = order_items["price"].quantile(0.75)

IQR_price = Q3_price - Q1_price

limite_superior_price = Q3_price + 1.5 * IQR_price

print(f"Q1: R$ {Q1_price:.2f}")
print(f"Q3: R$ {Q3_price:.2f}")
print(f"IQR: R$ {IQR_price:.2f}")
print(f"Limite superior: R$ {limite_superior_price:.2f}")

Q1: R$ 39.90
Q3: R$ 134.90
IQR: R$ 95.00
Limite superior: R$ 277.40


In [72]:
outliers_price = order_items[
    order_items["price"] > limite_superior_price
]

print(f"Quantidade de potenciais outliers: {len(outliers_price)}")

Quantidade de potenciais outliers: 8427


In [73]:
print(
    f"Percentual de potenciais outliers: "
    f"{len(outliers_price) / len(order_items) * 100:.2f}%"
)

Percentual de potenciais outliers: 7.48%


In [74]:
outliers_price[
    ["order_id", "product_id", "seller_id", "price", "freight_value"]
].sort_values("price", ascending=False).head(20)

,order_id,product_id,seller_id,price,freight_value
3556,0812eb902a67711a1cb742b3cdaa65ae,489ae2aa008f021502940f251d4cce7f,e3b4998c7a498169dc7bce44e6bb6277,6735.00,194.31
112233,fefacc66af859508bf1a7934eab1e97f,69c590f7ffc7bf8db97190b6cb6ed62e,80ceebb4ee9b31afb6c6a916a574a1e2,6729.00,193.21
107841,f5136e38d1a14a4dbd87dff67da82701,1bdf5e6731585cf01aa8169c7028d6ad,ee27a8f15b1dded4d213a468ba4eb391,6499.00,227.66
74336,a96610ab360d42a2e5335a3998b4718a,a6492cc69376c469ab6f61d8f44de961,59417c56835dd8e2e72f91f809cd4092,4799.00,151.34
11249,199af31afc78c699f0dbf71fb178d4d4,c3ed642d592594bb648ff4a04cee2747,59417c56835dd8e2e72f91f809cd4092,4690.00,74.34
62086,8dbc85d1447242f3b127dda390d56e19,259037a6a41845e455183f89c5035f18,c72de06d72748d1a0dfb2125be43ba63,4590.00,91.78
29193,426a9742b533fc6fed17d1fd6d143d7e,a1beef8f3992dbd4cd8726796aa69c53,512d298ac2a96d1931b6bd30aa21f61d,4399.87,113.45
45843,68101694e5c5dc7330c91e1bbc36214f,6cdf8fc1d741c76586d8b6b15e9eef30,ed4acab38528488b65a9a9c603ff024a,4099.99,75.27
78310,b239ca7cd485940b31882363b52e6674,dd113cb02b2af9c8e5787e8f1f0722f6,821fb029fc6e495ca4f08a35d51e53a5,4059.00,104.51
59137,86c4eab1571921a6a6e248ed312f5a5a,6902c1962dd19d540807d0ab8fade5c6,fa1c13f2614d7b5c4749cbc52fecda94,3999.90,17.01


In [75]:
outliers_price["product_id"].nunique()

3715

In [76]:
order_items[
    order_items["freight_value"] == 0
]["price"].describe()

count    383.000000
mean      98.601488
std       50.004247
min       53.900000
25%       69.900000
50%       99.900000
75%      106.900000
max      712.900000
Name: price, dtype: float64

In [77]:
order_items[
    order_items["freight_value"] == 0
]["order_id"].nunique()

339

In [78]:
order_items[
    order_items["freight_value"] == 0
]["product_id"].nunique()

10

In [79]:
freight_zero_products = (
    order_items[
        order_items["freight_value"] == 0
    ]["product_id"]
    .value_counts()
)

freight_zero_products

product_id
53b36df67ebb7c41585e8d54d6772e08    187
aca2eb7d00ea1a7b8ebd4e68314663af     98
422879e10f46682990de24d770e7f83d     56
7a10781637204d8d10485c71a6108a2e     27
f1c7f353075ce59d8a6f3cf58f419c9c      9
5a848e4ab52fd5445cdc07aab1c40e48      2
81fe540cb0119e1d4ef5f191701b3cb9      1
2b4609f8948be18874494203496bc318      1
2a34e0af5f72ca6cdeb148377a247c86      1
4fcb3d9a5f4871e8362dfedbdb02b064      1
Name: count, dtype: int64

In [80]:
order_items[
    (order_items["freight_value"] == 0)
][
    ["product_id", "price", "freight_value"]
].sort_values("price", ascending=False).head(20)

,product_id,price,freight_value
89654,2a34e0af5f72ca6cdeb148377a247c86,712.9,0.0
111497,7a10781637204d8d10485c71a6108a2e,219.0,0.0
99481,7a10781637204d8d10485c71a6108a2e,209.0,0.0
12028,7a10781637204d8d10485c71a6108a2e,209.0,0.0
88163,7a10781637204d8d10485c71a6108a2e,209.0,0.0
19489,7a10781637204d8d10485c71a6108a2e,199.0,0.0
94244,7a10781637204d8d10485c71a6108a2e,199.0,0.0
21865,7a10781637204d8d10485c71a6108a2e,199.0,0.0
20682,7a10781637204d8d10485c71a6108a2e,199.0,0.0
97216,7a10781637204d8d10485c71a6108a2e,199.0,0.0


In [81]:
order_items[
    order_items["product_id"].isin(freight_zero_products.index)
].groupby("product_id")["freight_value"].agg(
    ["count", "mean", "min", "max"]
)

,count,mean,min,max
product_id,,,,
2a34e0af5f72ca6cdeb148377a247c86,3,132.926667,0.0,256.13
2b4609f8948be18874494203496bc318,260,15.163538,0.0,185.73
422879e10f46682990de24d770e7f83d,484,15.752149,0.0,83.27
4fcb3d9a5f4871e8362dfedbdb02b064,89,21.236517,0.0,44.49
53b36df67ebb7c41585e8d54d6772e08,323,7.041827,0.0,51.31
5a848e4ab52fd5445cdc07aab1c40e48,197,15.517868,0.0,51.36
7a10781637204d8d10485c71a6108a2e,143,16.246434,0.0,68.45
81fe540cb0119e1d4ef5f191701b3cb9,2,4.670000,0.0,9.34
aca2eb7d00ea1a7b8ebd4e68314663af,527,13.684744,0.0,35.85


In [82]:
order_items[
    order_items["shipping_limit_date"].isna()
]["order_id"].count()

np.int64(0)

Auditoria da order_items

A tabela order_items possui 112.650 registros e não apresenta valores ausentes ou duplicações completas. A combinação order_id + order_item_id também não apresenta duplicidades, confirmando a consistência da identificação dos itens dentro dos pedidos.

A variável price apresentou distribuição assimétrica e 8.427 potenciais outliers pelo critério do IQR (7,48% dos registros). Esses valores não foram removidos, pois podem representar produtos legítimos de maior valor.

Foram identificados 383 registros com freight_value = 0, concentrados em 10 produtos. Como esses mesmos produtos apresentam também valores positivos de frete, os zeros não foram considerados erros e foram mantidos.

A variável shipping_limit_date não apresentou valores ausentes e foi convertida para o tipo datetime para análises temporais posteriores

In [12]:
customers = pd.read_csv(
    "../data/raw/olist_customers_dataset.csv"
)

In [84]:
customers.head()

,customer_id,customer_unique_id,customer_zip_code_prefix,customer_city,customer_state
0,06b8999e2fba1a1fbc88172c00ba8bc7,861eff4711a542e4b93843c6dd7febb0,14409,franca,SP
1,18955e83d337fd6b2def6b18a428ac77,290c77bc529b7ac935b93aa66c333dc3,9790,sao bernardo do campo,SP
2,4e7b3e00288586ebd08712fdd0374a03,060e732b5b29e8181a18229c7b0b2b5e,1151,sao paulo,SP
3,b2b6027bc5c5109e529d4dc6358b12c3,259dac757896d24d7702b9acbbff3f3c,8775,mogi das cruzes,SP
4,4f2d8ab171c80ec8364f7c12e35b23ad,345ecd01c38d18a9036ed96c73b8d066,13056,campinas,SP


In [85]:
customers.shape

(99441, 5)

In [86]:
customers.columns

Index(['customer_id', 'customer_unique_id', 'customer_zip_code_prefix',
       'customer_city', 'customer_state'],
      dtype='str')

In [87]:
customers.info()

<class 'pandas.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 5 columns):
 #   Column                    Non-Null Count  Dtype
---  ------                    --------------  -----
 0   customer_id               99441 non-null  str  
 1   customer_unique_id        99441 non-null  str  
 2   customer_zip_code_prefix  99441 non-null  int64
 3   customer_city             99441 non-null  str  
 4   customer_state            99441 non-null  str  
dtypes: int64(1), str(4)
memory usage: 3.8 MB


In [88]:
customers.isnull().sum()

customer_id                 0
customer_unique_id          0
customer_zip_code_prefix    0
customer_city               0
customer_state              0
dtype: int64

In [89]:
customers.isnull().mean() * 100

customer_id                 0.0
customer_unique_id          0.0
customer_zip_code_prefix    0.0
customer_city               0.0
customer_state              0.0
dtype: float64

In [90]:
customers.duplicated().sum()

np.int64(0)

In [91]:
customers["customer_id"].duplicated().sum()

np.int64(0)

In [92]:
customers["customer_id"].nunique()

99441

In [93]:
customers.shape[0]

99441

In [94]:
customers["customer_state"].value_counts()

customer_state
SP    41746
RJ    12852
MG    11635
RS     5466
PR     5045
SC     3637
BA     3380
DF     2140
ES     2033
GO     2020
PE     1652
CE     1336
PA      975
MT      907
MA      747
MS      715
PB      536
PI      495
RN      485
AL      413
SE      350
TO      280
RO      253
AM      148
AC       81
AP       68
RR       46
Name: count, dtype: int64

In [95]:
customers["customer_city"].nunique()

4119

In [96]:
customers["customer_unique_id"].duplicated().sum()

np.int64(3345)

In [97]:
customers["customer_unique_id"].nunique()

96096

In [98]:
customers[
    customers["customer_unique_id"].duplicated(keep=False)
].sort_values("customer_unique_id").head(20)

,customer_id,customer_unique_id,customer_zip_code_prefix,customer_city,customer_state
35608,24b0e2bd287e47d54d193e7bbb51103f,00172711b30d52eea8b313a7f2cced02,45200,jequie,BA
19299,1afe8a9c67eec3516c09a8bdcc539090,00172711b30d52eea8b313a7f2cced02,45200,jequie,BA
20023,1b4a75b3478138e99902678254b260f4,004288347e5e88a27ded2bb23747066c,26220,nova iguacu,RJ
22066,f6efe5d5c7b85e12355f9d5c3db46da2,004288347e5e88a27ded2bb23747066c,26220,nova iguacu,RJ
72451,49cf243e0d353cd418ca77868e24a670,004b45ec5c64187465168251cd1c9c2f,57055,maceio,AL
87012,d95f60d70d9ea9a7fe37c53c931940bb,004b45ec5c64187465168251cd1c9c2f,57035,maceio,AL
36269,8ac44e9c15d396b8c3c7cbab0fff4536,0058f300f57d7b93c477a131a59b36c3,41370,salvador,BA
61403,f530197ea86ced9488a03d055e118ebf,0058f300f57d7b93c477a131a59b36c3,40731,salvador,BA
87414,cbb68c721ba9ddb30d8a490cc1897fa1,00a39521eb40f7012db50455bf083460,72595,brasilia,DF
54881,876356df457f952458a764348e1858bc,00a39521eb40f7012db50455bf083460,72595,brasilia,DF


In [99]:
customers[
    customers["customer_unique_id"].duplicated(keep=False)
].groupby("customer_unique_id").agg(
    quantidade_customer_ids=("customer_id", "nunique"),
    quantidade_cidades=("customer_city", "nunique"),
    quantidade_estados=("customer_state", "nunique"),
    quantidade_ceps=("customer_zip_code_prefix", "nunique")
).describe()

,quantidade_customer_ids,quantidade_cidades,quantidade_estados,quantidade_ceps
count,2997.000000,2997.000000,2997.000000,2997.000000
mean,2.116116,1.041041,1.013347,1.084751
std,0.516610,0.200093,0.117646,0.283310
min,2.000000,1.000000,1.000000,1.000000
25%,2.000000,1.000000,1.000000,1.000000
50%,2.000000,1.000000,1.000000,1.000000
75%,2.000000,1.000000,1.000000,1.000000
max,17.000000,3.000000,3.000000,3.000000


In [16]:
order_payments = pd.read_csv(
    "../data/raw/olist_order_payments_dataset.csv"
)

In [103]:
order_payments.shape

(103886, 5)

In [104]:
order_payments.columns

Index(['order_id', 'payment_sequential', 'payment_type',
       'payment_installments', 'payment_value'],
      dtype='str')

In [105]:
order_payments.head()

,order_id,payment_sequential,payment_type,payment_installments,payment_value
0,b81ef226f3fe1789b1e8b2acac839d17,1,credit_card,8,99.33
1,a9810da82917af2d9aefd1278f1dcfa0,1,credit_card,1,24.39
2,25e8ea4e93396b6fa0d3dd708e76c1bd,1,credit_card,1,65.71
3,ba78997921bbcdc1373bb41e913ab953,1,credit_card,8,107.78
4,42fdf880ba16b47b59251dd489d4441a,1,credit_card,2,128.45


In [106]:
order_payments.info()

<class 'pandas.DataFrame'>
RangeIndex: 103886 entries, 0 to 103885
Data columns (total 5 columns):
 #   Column                Non-Null Count   Dtype  
---  ------                --------------   -----  
 0   order_id              103886 non-null  str    
 1   payment_sequential    103886 non-null  int64  
 2   payment_type          103886 non-null  str    
 3   payment_installments  103886 non-null  int64  
 4   payment_value         103886 non-null  float64
dtypes: float64(1), int64(2), str(2)
memory usage: 4.0 MB


In [107]:
order_payments.isnull().sum()

order_id                0
payment_sequential      0
payment_type            0
payment_installments    0
payment_value           0
dtype: int64

In [108]:
order_payments.isnull().mean() * 100

order_id                0.0
payment_sequential      0.0
payment_type            0.0
payment_installments    0.0
payment_value           0.0
dtype: float64

In [109]:
order_payments.duplicated().sum()

np.int64(0)

In [110]:
order_payments["order_id"].duplicated().sum()

np.int64(4446)

In [111]:
order_payments["order_id"].nunique()

99440

In [112]:
order_payments[
    ["order_id", "payment_sequential"]
].duplicated().sum()

np.int64(0)

In [113]:
order_payments[
    ["payment_value"]
].describe()

,payment_value
count,103886.000000
mean,154.100380
std,217.494064
min,0.000000
25%,56.790000
50%,100.000000
75%,171.837500
max,13664.080000


In [114]:
order_payments[
    ["payment_type"]
].value_counts()

payment_type
credit_card     76795
boleto          19784
voucher          5775
debit_card       1529
not_defined         3
Name: count, dtype: int64

In [115]:
order_payments[
    ["payment_installments"]
].describe()

,payment_installments
count,103886.000000
mean,2.853349
std,2.687051
min,0.000000
25%,1.000000
50%,1.000000
75%,4.000000
max,24.000000


In [116]:
order_payments["payment_installments"].value_counts().sort_index()

payment_installments
0         2
1     52546
2     12413
3     10461
4      7098
5      5239
6      3920
7      1626
8      4268
9       644
10     5328
11       23
12      133
13       16
14       15
15       74
16        5
17        8
18       27
20       17
21        3
22        1
23        1
24       18
Name: count, dtype: int64

In [117]:
order_payments[
    order_payments["payment_value"] == 0
]

,order_id,payment_sequential,payment_type,payment_installments,payment_value
19922,8bcbe01d44d147f901cd3192671144db,4,voucher,1,0.0
36822,fa65dad1b0e818e3ccc5cb0e39231352,14,voucher,1,0.0
43744,6ccb433e00daae1283ccc956189c82ae,4,voucher,1,0.0
51280,4637ca194b6387e2d538dc89b124b0ee,1,not_defined,1,0.0
57411,00b1cb0320190ca0daa2c88b35206009,1,not_defined,1,0.0
62674,45ed6e85398a87c253db47c2d9f48216,3,voucher,1,0.0
77885,fa65dad1b0e818e3ccc5cb0e39231352,13,voucher,1,0.0
94427,c8c528189310eaa44a745b8d9d26908b,1,not_defined,1,0.0
100766,b23878b3e8eb4d25a158f57d96331b18,4,voucher,1,0.0


In [118]:
order_payments[
    order_payments["payment_value"] == 0
][
    [
        "order_id",
        "payment_sequential",
        "payment_type",
        "payment_installments",
        "payment_value"
    ]
]

,order_id,payment_sequential,payment_type,payment_installments,payment_value
19922,8bcbe01d44d147f901cd3192671144db,4,voucher,1,0.0
36822,fa65dad1b0e818e3ccc5cb0e39231352,14,voucher,1,0.0
43744,6ccb433e00daae1283ccc956189c82ae,4,voucher,1,0.0
51280,4637ca194b6387e2d538dc89b124b0ee,1,not_defined,1,0.0
57411,00b1cb0320190ca0daa2c88b35206009,1,not_defined,1,0.0
62674,45ed6e85398a87c253db47c2d9f48216,3,voucher,1,0.0
77885,fa65dad1b0e818e3ccc5cb0e39231352,13,voucher,1,0.0
94427,c8c528189310eaa44a745b8d9d26908b,1,not_defined,1,0.0
100766,b23878b3e8eb4d25a158f57d96331b18,4,voucher,1,0.0


In [119]:
order_payments_not_defined = order_payments[
    order_payments["payment_type"] == "not_defined"
]

order_payments_not_defined

,order_id,payment_sequential,payment_type,payment_installments,payment_value
51280,4637ca194b6387e2d538dc89b124b0ee,1,not_defined,1,0.0
57411,00b1cb0320190ca0daa2c88b35206009,1,not_defined,1,0.0
94427,c8c528189310eaa44a745b8d9d26908b,1,not_defined,1,0.0


In [120]:
order_payments_not_defined["payment_value"].describe()

count    3.0
mean     0.0
std      0.0
min      0.0
25%      0.0
50%      0.0
75%      0.0
max      0.0
Name: payment_value, dtype: float64

In [121]:
order_payments_not_defined[
    ["payment_sequential", "payment_installments", "payment_value"]
]

,payment_sequential,payment_installments,payment_value
51280,1,1,0.0
57411,1,1,0.0
94427,1,1,0.0


In [123]:
payments_per_order = order_payments.groupby("order_id").size()

payments_per_order.describe()

count    99440.000000
mean         1.044710
std          0.381166
min          1.000000
25%          1.000000
50%          1.000000
75%          1.000000
max         29.000000
dtype: float64

In [124]:
payments_per_order.value_counts().sort_index()

1     96479
2      2382
3       301
4       108
5        52
6        36
7        28
8        11
9         9
10        5
11        8
12        8
13        3
14        2
15        2
19        2
21        1
22        1
26        1
29        1
Name: count, dtype: int64

In [125]:
payments_per_order[payments_per_order >= 10].sort_values(ascending=False)

order_id
fa65dad1b0e818e3ccc5cb0e39231352    29
ccf804e764ed5650cd8759557269dc13    26
285c2e15bebd4ac83635ccc563dc71f4    22
895ab968e7bb0d5659d16cd74cd1650c    21
ee9ca989fc93ba09a6eddc250ce01742    19
fedcd9f7ccdc8cba3a18defedd1a5547    19
4bfcba9e084f46c8e3cb49b0fa6e6159    15
21577126c19bf11a0b91592e5844ba78    15
4689b1816de42507a7d63a4617383c59    14
3c58bffb70dcf45f12bdf66a3c215905    14
4fb76fa13b108a0d0478483421b0992c    13
cf101c3abd3c061ca9f78c1bbb1125af    13
73df5d6adbeea12c8ae03df93f346e86    13
1a611328643ae11146ba09a4425d2e12    12
67d83bd36ec2c7fb557742fb58837659    12
68986e4324f6a21481df4e6e89abcf01    12
1d9a9731b9c10fc9cba74e6f74782e8b    12
465c2e1bee4561cb39e0db8c5993aafc    12
c6492b842ac190db807c15aff21a7dd6    12
d744783ed2ace06cac647a9e64dcbcfd    12
6d58638e32674bebee793a47ac4cbadc    12
0bbb3f7791a87d0307555e57da3a1ff1    11
370e2e6c1a9fd451eb7f0852daa3b006    11
27a940efdd448db29463b53ea0cfa2f4    11
4069c489933782af79afcd3a0e4d693c    11
1be51feefcd481be

In [126]:
orders_many_payments = payments_per_order[
    payments_per_order >= 10
].index

order_payments[
    order_payments["order_id"].isin(orders_many_payments)
].sort_values(
    ["order_id", "payment_sequential"]
)

,order_id,payment_sequential,payment_type,payment_installments,payment_value
92290,0bbb3f7791a87d0307555e57da3a1ff1,1,credit_card,7,76.20
23387,0bbb3f7791a87d0307555e57da3a1ff1,2,voucher,1,2.41
31570,0bbb3f7791a87d0307555e57da3a1ff1,3,voucher,1,2.39
49652,0bbb3f7791a87d0307555e57da3a1ff1,4,voucher,1,2.58
71125,0bbb3f7791a87d0307555e57da3a1ff1,5,voucher,1,2.58
...,...,...,...,...,...
7548,fedcd9f7ccdc8cba3a18defedd1a5547,15,voucher,1,10.66
64520,fedcd9f7ccdc8cba3a18defedd1a5547,16,voucher,1,9.54
48288,fedcd9f7ccdc8cba3a18defedd1a5547,17,voucher,1,7.51
87292,fedcd9f7ccdc8cba3a18defedd1a5547,18,voucher,1,11.12


In [127]:
order_payments[
    order_payments["order_id"].isin(orders_many_payments)
]["payment_type"].value_counts()

payment_type
voucher        448
credit_card     19
Name: count, dtype: int64

In [128]:
order_payments[
    order_payments["order_id"].isin(orders_many_payments)
].groupby("payment_type")["payment_value"].agg(
    ["count", "mean", "min", "max"]
)

,count,mean,min,max
payment_type,,,,
credit_card,19,40.230000,0.26,455.86
voucher,448,13.188259,0.00,150.00


In [129]:
order_payments[
    (order_payments["payment_type"] == "voucher") &
    (order_payments["payment_value"] == 0)
]

,order_id,payment_sequential,payment_type,payment_installments,payment_value
19922,8bcbe01d44d147f901cd3192671144db,4,voucher,1,0.0
36822,fa65dad1b0e818e3ccc5cb0e39231352,14,voucher,1,0.0
43744,6ccb433e00daae1283ccc956189c82ae,4,voucher,1,0.0
62674,45ed6e85398a87c253db47c2d9f48216,3,voucher,1,0.0
77885,fa65dad1b0e818e3ccc5cb0e39231352,13,voucher,1,0.0
100766,b23878b3e8eb4d25a158f57d96331b18,4,voucher,1,0.0


In [130]:
order_payments[
    (order_payments["payment_type"] == "voucher") &
    (order_payments["payment_value"] == 0)
]["order_id"].isin(orders_many_payments).sum()

np.int64(2)

In [131]:
order_payments[
    order_payments["payment_installments"] == 0
]

,order_id,payment_sequential,payment_type,payment_installments,payment_value
46982,744bade1fcf9ff3f31d860ace076d422,2,credit_card,0,58.69
79014,1a57108394169c0b47d8f876acc9ba2d,2,credit_card,0,129.94


In [132]:
order_payments[
    order_payments["payment_installments"] == 0
]["payment_type"].value_counts()

payment_type
credit_card    2
Name: count, dtype: int64

In [133]:
order_payments[
    order_payments["payment_installments"] == 0
]["payment_value"].describe()

count      2.000000
mean      94.315000
std       50.381358
min       58.690000
25%       76.502500
50%       94.315000
75%      112.127500
max      129.940000
Name: payment_value, dtype: float64

In [134]:
orders.columns

Index(['order_id', 'customer_id', 'order_status', 'order_purchase_timestamp',
       'order_approved_at', 'order_delivered_carrier_date',
       'order_delivered_customer_date', 'order_estimated_delivery_date'],
      dtype='str')

In [17]:
order_items_by_order = (
    order_items
    .groupby("order_id")
    .agg(
        total_price=("price", "sum"),
        total_freight=("freight_value", "sum")
    )
)

order_items_by_order.head()

NameError: name 'order_items' is not defined

In [14]:
order_items_by_order["total_order_value"] = (
    order_items_by_order["total_price"] +
    order_items_by_order["total_freight"]
)

order_items_by_order.head()

NameError: name 'order_items_by_order' is not defined

In [137]:
payments_by_order = (
    order_payments
    .groupby("order_id")
    .agg(
        total_payment_value=("payment_value", "sum")
    )
)

payments_by_order.head()

,total_payment_value
order_id,
00010242fe8c5a6d1ba2dd792cb16214,72.19
00018f77f2f0320c557190d7a144bdd3,259.83
000229ec398224ef6ca0657da4fc703e,216.87
00024acbcdf0a6daa1e931b038114c75,25.78
00042b26cf59d7ce69dfabb4e55b4fd9,218.04


In [138]:
order_financial_check = order_items_by_order.join(
    payments_by_order,
    how="left"
)

order_financial_check.head()

,total_price,total_freight,total_order_value,total_payment_value
order_id,,,,
00010242fe8c5a6d1ba2dd792cb16214,58.90,13.29,72.19,72.19
00018f77f2f0320c557190d7a144bdd3,239.90,19.93,259.83,259.83
000229ec398224ef6ca0657da4fc703e,199.00,17.87,216.87,216.87
00024acbcdf0a6daa1e931b038114c75,12.99,12.79,25.78,25.78
00042b26cf59d7ce69dfabb4e55b4fd9,199.90,18.14,218.04,218.04


In [143]:
order_financial_check["payment_difference"] = (
    order_financial_check["total_payment_value"]
    - order_financial_check["total_order_value"]
)

In [140]:
order_financial_check["payment_difference"].describe()

count    98665.000000
mean         0.029092
std          1.129221
min        -51.620000
25%          0.000000
50%          0.000000
75%          0.000000
max        182.810000
Name: payment_difference, dtype: float64

In [141]:
(order_financial_check["payment_difference"] == 0).sum()

np.int64(79051)

In [142]:
(order_financial_check["payment_difference"] != 0).sum()

np.int64(19615)

In [144]:
order_financial_check[
    order_financial_check["payment_difference"] != 0
]["payment_difference"].describe()

count    1.961400e+04
mean     1.463439e-01
std      2.529326e+00
min     -5.162000e+01
25%     -1.421085e-14
50%      3.552714e-15
75%      1.421085e-14
max      1.828100e+02
Name: payment_difference, dtype: float64

In [145]:
order_financial_check[
    order_financial_check["payment_difference"] != 0
]["payment_difference"].quantile(
    [0.01, 0.05, 0.25, 0.50, 0.75, 0.95, 0.99]
)

0.01   -1.000000e-02
0.05   -5.684342e-14
0.25   -1.421085e-14
0.50    3.552714e-15
0.75    1.421085e-14
0.95    5.684342e-14
0.99    2.914800e+00
Name: payment_difference, dtype: float64

In [146]:
tolerance = 0.01

real_payment_divergences = order_financial_check[
    order_financial_check["payment_difference"].abs() > tolerance
]

In [147]:
print(f"Quantidade de divergências reais: {len(real_payment_divergences)}")

Quantidade de divergências reais: 381


In [148]:
print(
    f"Percentual de divergências reais: "
    f"{len(real_payment_divergences) / len(order_financial_check) * 100:.2f}%"
)

Percentual de divergências reais: 0.39%


In [149]:
real_payment_divergences["payment_difference"].describe()

count    381.000000
mean       7.534961
std       16.564267
min      -51.620000
25%        0.010000
50%        3.090000
75%        9.570000
max      182.810000
Name: payment_difference, dtype: float64

In [150]:
real_payment_divergences[
    real_payment_divergences["payment_difference"] < 0
].shape

(90, 5)

In [152]:
real_payment_divergences[
    real_payment_divergences["payment_difference"] > 0
].shape

(291, 5)

In [156]:
negative_divergences = real_payment_divergences[
    real_payment_divergences["payment_difference"] < 0
]

positive_divergences = real_payment_divergences[
    real_payment_divergences["payment_difference"] > 0
]

In [154]:
negative_divergences["payment_difference"].describe()

count    90.000000
mean     -2.217667
std       6.784611
min     -51.620000
25%      -0.027500
50%      -0.010000
75%      -0.010000
max      -0.010000
Name: payment_difference, dtype: float64

In [155]:
positive_divergences["payment_difference"].describe()

count    291.000000
mean      10.551237
std       17.514328
min        0.010000
25%        1.875000
50%        5.550000
75%       12.140000
max      182.810000
Name: payment_difference, dtype: float64

In [157]:
positive_payment_counts = (
    order_payments[
        order_payments["order_id"].isin(positive_divergences.index)
    ]
    .groupby("order_id")
    .size()
)

positive_payment_counts.describe()

count    291.000000
mean       1.065292
std        0.247466
min        1.000000
25%        1.000000
50%        1.000000
75%        1.000000
max        2.000000
dtype: float64

In [158]:
positive_payment_counts.value_counts().sort_index()

1    272
2     19
Name: count, dtype: int64

In [159]:
positive_payment_types = (
    order_payments[
        order_payments["order_id"].isin(positive_divergences.index)
    ]
    .groupby("order_id")["payment_type"]
    .agg(lambda x: ", ".join(x))
)

positive_payment_types.value_counts()

payment_type
credit_card                 255
boleto                       16
credit_card, credit_card     11
credit_card, voucher          6
voucher, credit_card          2
debit_card                    1
Name: count, dtype: int64

In [160]:
positive_divergences[
    [
        "total_price",
        "total_freight",
        "total_order_value",
        "total_payment_value",
        "payment_difference"
    ]
].sort_values(
    "payment_difference",
    ascending=False
).head(20)

,total_price,total_freight,total_order_value,total_payment_value,payment_difference
order_id,,,,,
ce6d150fb29ada17d2082f4847107665,1299.00,104.66,1403.66,1586.47,182.81
6e5fe7366a2e1bfbf3257dba0af1267f,179.19,108.72,287.91,406.92,119.01
70b742795bc441e94a44a084b6d9ce7a,269.99,196.94,466.93,578.82,111.89
996c7e73600ad3723e8627ab7bef81e4,559.90,28.00,587.90,664.43,76.53
70b7e94ea46d3e8b5bc12a50186edaf0,167.88,45.27,213.15,274.84,61.69
bc2c82b0ef78d2252b6176d1972db7c9,165.00,77.01,242.01,303.02,61.01
af9ffff2ce6b3defd34fd4c78857a379,395.65,17.52,413.17,466.97,53.80
bfdb5bbb06458d600a33d61f5f287472,297.00,51.93,348.93,394.36,45.43
8d9c0dc8d5a2ce804f6b925d8f8e6c1d,209.80,44.65,254.45,293.89,39.44


In [161]:
order_payments[
    order_payments["order_id"].isin(
        positive_divergences.sort_values(
            "payment_difference",
            ascending=False
        ).head(20).index
    )
].sort_values(
    ["order_id", "payment_sequential"]
)

,order_id,payment_sequential,payment_type,payment_installments,payment_value
71389,32720c0c8b42f4c2cdb07cee5ec2444b,1,credit_card,12,259.59
17085,513a94a0164453b4777d125b836f7ca1,1,credit_card,12,240.47
93706,6b0ac1bdea322c7060eea92feb9e9a6f,1,credit_card,8,362.56
38060,6e5fe7366a2e1bfbf3257dba0af1267f,1,credit_card,10,406.92
25761,70b742795bc441e94a44a084b6d9ce7a,1,credit_card,20,578.82
2970,70b7e94ea46d3e8b5bc12a50186edaf0,1,credit_card,24,274.84
95180,8d9c0dc8d5a2ce804f6b925d8f8e6c1d,1,credit_card,12,293.89
61856,996c7e73600ad3723e8627ab7bef81e4,1,credit_card,10,664.43
3038,a69aaa894f707814637e9794aa556f0f,1,credit_card,15,211.87
22304,abf1130bc676c9dcadf91e24f5e30a30,1,credit_card,6,496.89


In [162]:
items_per_order = (
    order_items
    .groupby("order_id")
    .size()
)

positive_divergences["item_count"] = (
    positive_divergences.index.map(items_per_order)
)

positive_divergences["item_count"].describe()

count    291.000000
mean       1.728522
std        1.725622
min        1.000000
25%        1.000000
50%        1.000000
75%        1.000000
max       12.000000
Name: item_count, dtype: float64

In [163]:
positive_divergences["item_count"].value_counts().sort_index()

item_count
1     222
2      22
3      14
4      11
5       5
6       7
7       6
8       1
11      2
12      1
Name: count, dtype: int64

In [164]:
top_positive_orders = (
    positive_divergences
    .sort_values("payment_difference", ascending=False)
    .head(20)
    .index
)

orders[
    orders["order_id"].isin(top_positive_orders)
].sort_values("order_id")

,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date
72098,32720c0c8b42f4c2cdb07cee5ec2444b,8df8c71565b1b7c5d8bec24a7052ba2d,delivered,2017-11-24 22:46:55,2017-11-25 01:52:03,2017-11-28 18:41:04,2017-12-07 21:34:06,2017-12-22
81220,513a94a0164453b4777d125b836f7ca1,5c515f4bb8616f7feee867ad9a301bf9,delivered,2018-06-11 22:14:06,2018-06-11 22:41:00,2018-06-14 06:16:00,2018-06-21 21:08:39,2018-07-23
33384,6b0ac1bdea322c7060eea92feb9e9a6f,78d3ee0870e91f6ee67363695e0a705b,delivered,2018-06-28 14:29:37,2018-06-28 14:50:53,2018-06-29 14:15:00,2018-07-09 20:52:47,2018-08-01
48686,6e5fe7366a2e1bfbf3257dba0af1267f,5263febf1413e9a60ceb7b842008b7ef,delivered,2016-10-09 22:36:25,2016-10-10 12:02:19,2016-10-14 12:02:21,2016-10-17 13:02:21,2016-12-06
70865,70b742795bc441e94a44a084b6d9ce7a,0c155574f4f4d27594dbdd37731a6ecb,delivered,2017-08-24 10:29:36,2017-08-24 10:35:22,2017-08-30 15:28:21,2017-09-12 20:27:39,2017-09-20
52985,70b7e94ea46d3e8b5bc12a50186edaf0,1c4c6fd6cfc21396ae93680bfc5c34d4,delivered,2018-01-01 13:06:57,2018-01-01 13:17:25,2018-01-03 11:03:40,2018-01-10 00:24:12,2018-02-01
8548,8d9c0dc8d5a2ce804f6b925d8f8e6c1d,8c65f37a2cb3d0bedced14e105742bae,delivered,2018-03-19 13:40:14,2018-03-19 13:55:33,2018-03-20 00:28:42,2018-04-10 23:28:27,2018-04-10
33150,996c7e73600ad3723e8627ab7bef81e4,16be3ab0d37467d1f96d6fc30613ee57,delivered,2018-07-16 09:58:16,2018-07-16 10:10:18,2018-07-18 11:01:00,2018-07-20 22:11:39,2018-07-30
13942,a69aaa894f707814637e9794aa556f0f,ec27f1360f5b6f250aadf778bce052d7,delivered,2018-05-05 20:56:40,2018-05-05 21:10:36,2018-05-07 12:11:00,2018-05-11 13:09:03,2018-06-05
30187,abf1130bc676c9dcadf91e24f5e30a30,ec6bd30783896fce1908161c537f69e2,delivered,2018-01-11 16:24:13,2018-01-11 16:35:28,2018-01-12 21:36:10,2018-01-19 19:58:54,2018-02-05


In [165]:
top_positive_analysis = positive_divergences.loc[
    top_positive_orders,
    [
        "total_price",
        "total_freight",
        "total_order_value",
        "total_payment_value",
        "payment_difference"
    ]
].copy()

top_positive_analysis["payment_price_ratio"] = (
    top_positive_analysis["total_payment_value"]
    / top_positive_analysis["total_price"]
)

top_positive_analysis.sort_values(
    "payment_difference",
    ascending=False
)

,total_price,total_freight,total_order_value,total_payment_value,payment_difference,payment_price_ratio
order_id,,,,,,
ce6d150fb29ada17d2082f4847107665,1299.00,104.66,1403.66,1586.47,182.81,1.221301
6e5fe7366a2e1bfbf3257dba0af1267f,179.19,108.72,287.91,406.92,119.01,2.270886
70b742795bc441e94a44a084b6d9ce7a,269.99,196.94,466.93,578.82,111.89,2.143857
996c7e73600ad3723e8627ab7bef81e4,559.90,28.00,587.90,664.43,76.53,1.186694
70b7e94ea46d3e8b5bc12a50186edaf0,167.88,45.27,213.15,274.84,61.69,1.637122
bc2c82b0ef78d2252b6176d1972db7c9,165.00,77.01,242.01,303.02,61.01,1.836485
af9ffff2ce6b3defd34fd4c78857a379,395.65,17.52,413.17,466.97,53.80,1.180260
bfdb5bbb06458d600a33d61f5f287472,297.00,51.93,348.93,394.36,45.43,1.327811
8d9c0dc8d5a2ce804f6b925d8f8e6c1d,209.80,44.65,254.45,293.89,39.44,1.400810


In [13]:
order_items[
    order_items["order_id"].isin(
        top_positive_orders[:3]
    )
].sort_values(
    ["order_id", "order_item_id"]
)

NameError: name 'order_items' is not defined

In [167]:
top_3_orders = top_positive_orders[:3]

orders[
    orders["order_id"].isin(top_3_orders)
][
    [
        "order_id",
        "customer_id",
        "order_status",
        "order_purchase_timestamp",
        "order_approved_at"
    ]
]

,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at
11791,ce6d150fb29ada17d2082f4847107665,d8acb46c718975948423dbf1b603b959,delivered,2018-01-25 11:17:02,2018-01-27 12:35:38
48686,6e5fe7366a2e1bfbf3257dba0af1267f,5263febf1413e9a60ceb7b842008b7ef,delivered,2016-10-09 22:36:25,2016-10-10 12:02:19
70865,70b742795bc441e94a44a084b6d9ce7a,0c155574f4f4d27594dbdd37731a6ecb,delivered,2017-08-24 10:29:36,2017-08-24 10:35:22


In [168]:
order_payments[
    order_payments["order_id"].isin(top_3_orders)
].sort_values(
    ["order_id", "payment_sequential"]
)

,order_id,payment_sequential,payment_type,payment_installments,payment_value
38060,6e5fe7366a2e1bfbf3257dba0af1267f,1,credit_card,10,406.92
25761,70b742795bc441e94a44a084b6d9ce7a,1,credit_card,20,578.82
359,ce6d150fb29ada17d2082f4847107665,1,credit_card,10,1586.47


In [169]:
order_items[
    order_items["order_id"].isin(top_3_orders)
].groupby("order_id").agg(
    quantidade_itens=("order_item_id", "count"),
    total_price=("price", "sum"),
    total_freight=("freight_value", "sum")
)

,quantidade_itens,total_price,total_freight
order_id,,,
6e5fe7366a2e1bfbf3257dba0af1267f,6,179.19,108.72
70b742795bc441e94a44a084b6d9ce7a,1,269.99,196.94
ce6d150fb29ada17d2082f4847107665,1,1299.00,104.66


In [170]:
positive_divergences_with_payment = (
    positive_divergences
    .join(
        order_payments[
            [
                "order_id",
                "payment_installments"
            ]
        ].set_index("order_id"),
        how="left"
    )
)

In [171]:
positive_divergences_with_payment[
    ["payment_difference", "payment_installments"]
].corr()

,payment_difference,payment_installments
payment_difference,1.000000,0.609928
payment_installments,0.609928,1.000000


In [172]:
positive_divergences_with_payment.groupby(
    "payment_installments"
)["payment_difference"].agg(
    ["count", "mean", "median", "max"]
).sort_index()

,count,mean,median,max
payment_installments,,,,
1,45,0.696000,0.020,9.26
2,17,0.925882,1.010,2.11
3,40,2.992000,2.400,14.54
4,39,3.754103,3.640,8.74
5,29,6.762759,5.830,20.78
6,32,8.861563,7.230,38.09
7,18,10.048333,8.725,23.89
8,18,13.351111,11.150,34.69
9,7,15.485714,12.170,34.62


In [176]:
positive_divergences_with_payment = (
    positive_divergences
    .join(
        order_payments[
            [
                "order_id",
                "payment_type",
                "payment_installments"
            ]
        ].set_index("order_id"),
        how="left"
    )
)

In [177]:
positive_divergences_with_payment.columns

Index(['total_price', 'total_freight', 'total_order_value',
       'total_payment_value', 'payment_difference', 'item_count',
       'payment_type', 'payment_installments'],
      dtype='str')

In [178]:
credit_card_divergences = positive_divergences_with_payment[
    positive_divergences_with_payment["payment_type"] == "credit_card"
]

In [179]:
credit_card_divergences.groupby(
    "payment_installments"
)["payment_difference"].agg(
    ["count", "mean", "median", "max"]
).sort_index()

,count,mean,median,max
payment_installments,,,,
1,20,0.678000,0.355,4.23
2,17,0.925882,1.010,2.11
3,40,2.992000,2.400,14.54
4,39,3.754103,3.640,8.74
5,29,6.762759,5.830,20.78
6,32,8.861563,7.230,38.09
7,18,10.048333,8.725,23.89
8,18,13.351111,11.150,34.69
9,7,15.485714,12.170,34.62


In [180]:
credit_card_divergences[
    ["payment_difference", "payment_installments"]
].corr()

,payment_difference,payment_installments
payment_difference,1.000000,0.600297
payment_installments,0.600297,1.000000


In [181]:
print(
    f"Positivas: {len(positive_divergences)} "
    f"({len(positive_divergences) / len(real_payment_divergences) * 100:.2f}%)"
)

print(
    f"Negativas: {len(negative_divergences)} "
    f"({len(negative_divergences) / len(real_payment_divergences) * 100:.2f}%)"
)

Positivas: 291 (76.38%)
Negativas: 90 (23.62%)


In [182]:
print(
    f"Divergências reais: {len(real_payment_divergences)} "
    f"({len(real_payment_divergences) / len(order_financial_check) * 100:.2f}%)"
)

Divergências reais: 381 (0.39%)


# Auditoria Financeira — Pedidos × Pagamentos

## Objetivo

Verificar a consistência financeira dos pedidos, comparando o valor esperado da transação com o valor efetivamente registrado na tabela de pagamentos.

A análise considera como valor esperado do pedido:

**valor dos produtos + valor do frete**

Esse valor é comparado ao total dos pagamentos registrados para o mesmo `order_id`.

---

## 1. Construção do valor total do pedido

Os valores dos itens foram agregados por pedido:

- `total_price`: soma dos preços dos itens;
- `total_freight`: soma dos valores de frete;
- `total_order_value`: valor total esperado do pedido.

O cálculo utilizado foi:

`total_order_value = total_price + total_freight`

---

## 2. Construção do valor total pago

Os registros da tabela `order_payments` foram agrupados por `order_id`, somando todos os pagamentos associados a cada pedido.

Foi criada a variável:

`total_payment_value`

representando o valor total registrado como pago para cada pedido.

---

## 3. Comparação financeira

As duas informações foram relacionadas por `order_id`.

Foi criada a variável:

`payment_difference = total_payment_value - total_order_value`

Interpretação:

- `payment_difference = 0` → valores compatíveis;
- `payment_difference > 0` → pagamento superior ao valor calculado;
- `payment_difference < 0` → pagamento inferior ao valor calculado.

---

## 4. Tratamento de diferenças de ponto flutuante

Durante a comparação foram identificadas pequenas diferenças numéricas próximas de zero, compatíveis com limitações de representação de números decimais em operações computacionais.

Por isso foi adotada uma tolerância de:

`tolerance = 0.01`

Foram consideradas divergências reais apenas as observações em que:

`abs(payment_difference) > 0.01`

---

## 5. Resultado geral

Foram identificadas:

- **381 divergências financeiras reais**;
- **0,39%** dos pedidos analisados apresentaram divergências acima da tolerância.

Portanto, **99,61% dos pedidos analisados não apresentaram divergência financeira superior a R$ 0,01**.

---

## 6. Direção das divergências

Entre as 381 divergências reais:

- **291 (76,38%)** apresentaram pagamento superior ao valor calculado;
- **90 (23,62%)** apresentaram pagamento inferior ao valor calculado.

As divergências positivas foram, portanto, predominantes.

---

## 7. Magnitude das divergências

### Divergências negativas

Foram identificadas 90 ocorrências.

- média: aproximadamente **-R$ 2,22**;
- mediana: aproximadamente **-R$ 0,01**;
- mínimo: **-R$ 51,62**.

A mediana próxima de zero indica que grande parte dessas divergências é pequena, apesar da existência de alguns casos mais extremos.

### Divergências positivas

Foram identificadas 291 ocorrências.

- média: aproximadamente **R$ 10,55**;
- mediana: aproximadamente **R$ 5,55**;
- máximo: **R$ 182,81**.

As divergências positivas apresentam maior magnitude e merecem atenção especial nos casos extremos.

---

## 8. Investigação das divergências positivas

A análise dos meios de pagamento mostrou predominância de cartões de crédito entre as divergências positivas.

Também foram observados pedidos com múltiplos registros de pagamento, embora a maioria das divergências positivas esteja associada a apenas um registro de pagamento.

Entre as 291 divergências positivas:

- **272 pedidos** possuíam um registro de pagamento;
- **19 pedidos** possuíam dois registros de pagamento.

Nos casos analisados, o cartão de crédito foi o principal meio de pagamento associado às divergências positivas.

---

## 9. Relação com parcelamento

Foi analisada a relação entre `payment_difference` e `payment_installments`.

A correlação de Pearson observada foi aproximadamente:

`r = 0.60`

indicando uma **associação positiva moderada** entre número de parcelas e magnitude da divergência positiva.

Além disso, a análise agrupada mostrou tendência de aumento da divergência média conforme aumentava o número de parcelas.

Esse resultado indica que pedidos com maior quantidade de parcelas podem representar um grupo relevante para investigação.

Entretanto, **correlação não implica causalidade**. Portanto, não é possível concluir que o parcelamento seja a causa das divergências apenas com essa análise.

---

## 10. Casos extremos

Entre os maiores valores de divergência positiva foram encontrados pedidos com diferenças superiores a R$ 100.

O maior caso identificado apresentou:

- valor dos produtos: **R$ 1.299,00**;
- frete: **R$ 104,66**;
- valor esperado: **R$ 1.403,66**;
- valor registrado como pago: **R$ 1.586,47**;
- divergência: **R$ 182,81**.

A investigação individual desses casos mostrou que as maiores divergências não estavam necessariamente relacionadas à existência de múltiplos pagamentos.

Também foram observados pedidos com quantidade elevada de parcelas.

---

## 11. Conclusão da auditoria

A auditoria financeira indica que os registros de pedidos e pagamentos apresentam **alta consistência geral**, uma vez que apenas **0,39%** dos pedidos analisados apresentaram divergência superior à tolerância estabelecida de R$ 0,01.

Apesar da baixa proporção, as divergências não devem ser ignoradas, principalmente devido à existência de casos com valores elevados.

As principais evidências encontradas foram:

- predominância de divergências positivas;
- maior magnitude média das divergências positivas;
- predominância de cartão de crédito entre essas ocorrências;
- associação moderada entre número de parcelas e magnitude da divergência;
- existência de poucos casos extremos que merecem investigação individual.

### Classificação

**Status geral:** baixa incidência de inconsistências financeiras.

**Ponto de atenção:** casos extremos de divergência positiva e pedidos com maior número de parcelas.

**Próxima etapa:** investigar a consistência temporal dos pedidos, verificando a ordem lógica dos eventos entre compra, aprovação, envio e entrega.

In [6]:
orders[
    [
        "order_purchase_timestamp",
        "order_approved_at",
        "order_delivered_carrier_date",
        "order_delivered_customer_date",
        "order_estimated_delivery_date"
    ]
].dtypes

NameError: name 'orders' is not defined

In [5]:
date_columns = [
    "order_purchase_timestamp",
    "order_approved_at",
    "order_delivered_carrier_date",
    "order_delivered_customer_date",
    "order_estimated_delivery_date"
]

orders[date_columns] = orders[date_columns].apply(
    pd.to_datetime
)

NameError: name 'orders' is not defined

In [186]:
orders[date_columns].dtypes

order_purchase_timestamp         datetime64[us]
order_approved_at                datetime64[us]
order_delivered_carrier_date     datetime64[us]
order_delivered_customer_date    datetime64[us]
order_estimated_delivery_date    datetime64[us]
dtype: object

In [187]:
orders[date_columns].isna().sum()

order_purchase_timestamp            0
order_approved_at                 160
order_delivered_carrier_date     1783
order_delivered_customer_date    2965
order_estimated_delivery_date       0
dtype: int64

In [201]:
orders.groupby("order_status")[date_columns].apply(
    lambda x: x.isna().sum()
)

,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date
order_status,,,,,
approved,0,0,2,2,0
canceled,0,141,550,619,0
created,0,5,5,5,0
delivered,0,14,2,8,0
invoiced,0,0,314,314,0
processing,0,0,301,301,0
shipped,0,0,0,1107,0
unavailable,0,0,609,609,0


In [202]:
orders["purchase_after_approval"] = (
    orders["order_purchase_timestamp"]
    > orders["order_approved_at"]
)

In [190]:
orders["purchase_after_approval"].sum()

np.int64(0)

In [191]:
orders["approval_after_carrier"] = (
    orders["order_approved_at"]
    > orders["order_delivered_carrier_date"]
)

In [192]:
orders["carrier_after_customer_delivery"] = (
    orders["order_delivered_carrier_date"]
    > orders["order_delivered_customer_date"]
)

In [193]:
mask = (
    orders["order_purchase_timestamp"].notna()
    & orders["order_approved_at"].notna()
)

purchase_approval_anomalies = orders[
    mask &
    (
        orders["order_purchase_timestamp"]
        > orders["order_approved_at"]
    )
]

In [ ]:
orders["approval_time"] = (
    orders["order_approved_at"]
    - orders["order_purchase_timestamp"]
)

In [195]:
orders["carrier_time"] = (
    orders["order_delivered_carrier_date"]
    - orders["order_approved_at"]
)

In [203]:
orders["delivery_time"] = (
    orders["order_delivered_customer_date"]
    - orders["order_delivered_carrier_date"]
)

In [197]:
orders["approval_days"] = (
    orders["approval_time"].dt.total_seconds() / 86400
)

In [198]:
orders["delivery_delay_days"] = (
    orders["order_delivered_customer_date"]
    - orders["order_estimated_delivery_date"]
).dt.total_seconds() / 86400

In [199]:
orders["delivery_late"] = (
    orders["delivery_delay_days"] > 0
)

In [200]:
orders["delivery_late"].value_counts()

delivery_late
False    91614
True      7827
Name: count, dtype: int64

In [205]:
orders["delay_days"] = (
    orders["order_delivered_customer_date"]
    - orders["order_estimated_delivery_date"]
).dt.total_seconds() / (60 * 60 * 24)

In [206]:
orders.loc[
    orders["delivery_late"] == True,
    "delay_days"
].describe()

count    7827.000000
mean        9.552206
std        13.951700
min         0.002500
25%         1.867193
50%         5.807894
75%        11.822471
max       188.975081
Name: delay_days, dtype: float64

In [207]:
orders.loc[
    orders["delivery_late"] == True,
    "delay_days"
].quantile(
    [0.25, 0.50, 0.75, 0.90, 0.95, 0.99]
)

0.25     1.867193
0.50     5.807894
0.75    11.822471
0.90    21.531486
0.95    29.639211
0.99    62.759613
Name: delay_days, dtype: float64

In [12]:
orders[
    (orders["order_status"] == "delivered") &
    (orders["order_delivered_customer_date"].isna())
]

NameError: name 'orders' is not defined

In [209]:
delivered_missing_delivery = orders[
    (orders["order_status"] == "delivered") &
    (orders["order_delivered_customer_date"].isna())
]

delivered_missing_delivery[
    [
        "order_id",
        "order_purchase_timestamp",
        "order_approved_at",
        "order_delivered_carrier_date",
        "order_delivered_customer_date",
        "order_estimated_delivery_date"
    ]
]

,order_id,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date
3002,2d1e2d5bf4dc7227b3bfebb81328c15f,2017-11-28 17:44:07,2017-11-28 17:56:40,2017-11-30 18:12:23,NaT,2017-12-18
20618,f5dd62b788049ad9fc0526e3ad11a097,2018-06-20 06:58:43,2018-06-20 07:19:05,2018-06-25 08:05:00,NaT,2018-07-16
43834,2ebdfc4f15f23b91474edf87475f108e,2018-07-01 17:05:11,2018-07-01 17:15:12,2018-07-03 13:57:00,NaT,2018-07-30
79263,e69f75a717d64fc5ecdfae42b2e8e086,2018-07-01 22:05:55,2018-07-01 22:15:14,2018-07-03 13:57:00,NaT,2018-07-30
82868,0d3268bad9b086af767785e3f0fc0133,2018-07-01 21:14:02,2018-07-01 21:29:54,2018-07-03 09:28:00,NaT,2018-07-24
92643,2d858f451373b04fb5c984a1cc2defaf,2017-05-25 23:22:43,2017-05-25 23:30:16,NaT,NaT,2017-06-23
97647,ab7c89dc1bf4a1ead9d6ec1ec8968a84,2018-06-08 12:09:39,2018-06-08 12:36:39,2018-06-12 14:10:00,NaT,2018-06-26
98038,20edc82cf5400ce95e1afacc25798b31,2018-06-27 16:09:12,2018-06-27 16:29:30,2018-07-03 19:26:00,NaT,2018-07-19


In [210]:
delivered_missing_items = order_items[
    order_items["order_id"].isin(
        delivered_missing_delivery["order_id"]
    )
]

delivered_missing_items.groupby("order_id").agg(
    quantidade_itens=("order_item_id", "count"),
    total_price=("price", "sum"),
    total_freight=("freight_value", "sum")
)

,quantidade_itens,total_price,total_freight
order_id,,,
0d3268bad9b086af767785e3f0fc0133,1,188.99,15.63
20edc82cf5400ce95e1afacc25798b31,1,45.90,9.07
2d1e2d5bf4dc7227b3bfebb81328c15f,1,117.30,17.53
2d858f451373b04fb5c984a1cc2defaf,1,179.00,15.00
2ebdfc4f15f23b91474edf87475f108e,1,139.00,19.07
ab7c89dc1bf4a1ead9d6ec1ec8968a84,1,110.99,9.13
e69f75a717d64fc5ecdfae42b2e8e086,1,139.00,19.07
f5dd62b788049ad9fc0526e3ad11a097,1,329.00,25.24


Foram identificados 8 pedidos classificados como delivered sem registro da data de entrega ao cliente. A análise dos itens demonstra que os 8 pedidos possuem produtos associados, indicando que não se trata de pedidos sem movimentação comercial registrada. Entretanto, ainda é necessário verificar os registros de pagamento para avaliar a consistência financeira dessas operações.


In [211]:
delivered_missing_payments = order_payments[
    order_payments["order_id"].isin(
        delivered_missing_delivery["order_id"]
    )
]

delivered_missing_payments.sort_values(
    ["order_id", "payment_sequential"]
)

,order_id,payment_sequential,payment_type,payment_installments,payment_value
7536,0d3268bad9b086af767785e3f0fc0133,1,credit_card,4,204.62
10251,20edc82cf5400ce95e1afacc25798b31,1,credit_card,1,54.97
16949,2d1e2d5bf4dc7227b3bfebb81328c15f,1,credit_card,3,134.83
47287,2d858f451373b04fb5c984a1cc2defaf,1,credit_card,4,194.00
28032,2ebdfc4f15f23b91474edf87475f108e,1,credit_card,3,158.07
102958,ab7c89dc1bf4a1ead9d6ec1ec8968a84,1,credit_card,5,120.12
58017,e69f75a717d64fc5ecdfae42b2e8e086,1,credit_card,1,158.07
95941,f5dd62b788049ad9fc0526e3ad11a097,1,debit_card,1,354.24


In [212]:
delivered_missing_payments.groupby("order_id").agg(
    quantidade_pagamentos=("payment_sequential", "count"),
    total_pagamento=("payment_value", "sum")
)

,quantidade_pagamentos,total_pagamento
order_id,,
0d3268bad9b086af767785e3f0fc0133,1,204.62
20edc82cf5400ce95e1afacc25798b31,1,54.97
2d1e2d5bf4dc7227b3bfebb81328c15f,1,134.83
2d858f451373b04fb5c984a1cc2defaf,1,194.00
2ebdfc4f15f23b91474edf87475f108e,1,158.07
ab7c89dc1bf4a1ead9d6ec1ec8968a84,1,120.12
e69f75a717d64fc5ecdfae42b2e8e086,1,158.07
f5dd62b788049ad9fc0526e3ad11a097,1,354.24


Foram identificados 8 pedidos classificados como delivered sem registro da data de entrega ao cliente. Todos possuem itens e pagamentos registrados, indicando que não se trata de pedidos sem movimentação comercial. Em 7 casos, existe registro de entrega à transportadora, mas a data de entrega ao cliente está ausente; em 1 caso, a própria data de entrega à transportadora também está ausente. O achado caracteriza uma inconsistência entre o status operacional do pedido e sua trilha temporal, devendo ser investigado se decorre de falha de registro, integração ou atualização do status.

In [11]:
orders[
    [
        "order_purchase_timestamp",
        "order_approved_at",
        "order_delivered_carrier_date",
        "order_delivered_customer_date",
        "order_estimated_delivery_date"
    ]
].isna().sum()

NameError: name 'orders' is not defined

In [10]:
delivered_orders = orders[
    orders["order_status"] == "delivered"
].copy()

NameError: name 'orders' is not defined

In [215]:
len(delivered_orders)

96478

In [9]:
delivered_orders["order_delivered_customer_date"].isna().sum()

NameError: name 'delivered_orders' is not defined

In [3]:
temporal_analysis = delivered_orders.dropna(
    subset=[
        "order_purchase_timestamp",
        "order_approved_at",
        "order_delivered_carrier_date",
        "order_delivered_customer_date",
        "order_estimated_delivery_date"
    ]
).copy()

NameError: name 'delivered_orders' is not defined

In [231]:
len(temporal_analysis)

96455

In [219]:
temporal_analysis.isna().sum()

order_id                           0
customer_id                        0
order_status                       0
order_purchase_timestamp           0
order_approved_at                  0
order_delivered_carrier_date       0
order_delivered_customer_date      0
order_estimated_delivery_date      0
purchase_after_approval            0
approval_after_carrier             0
carrier_after_customer_delivery    0
approval_time                      0
carrier_time                       0
delivery_time                      0
approval_days                      0
delivery_delay_days                0
delivery_late                      0
delay_days                         0
dtype: int64

In [232]:
temporal_analysis["approval_time"] = (
    temporal_analysis["order_approved_at"]
    - temporal_analysis["order_purchase_timestamp"]
)

In [233]:
temporal_analysis["carrier_time"] = (
    temporal_analysis["order_delivered_carrier_date"]
    - temporal_analysis["order_approved_at"]
)

In [234]:
temporal_analysis["delivery_time"] = (
    temporal_analysis["order_delivered_customer_date"]
    - temporal_analysis["order_delivered_carrier_date"]
)

In [235]:
temporal_analysis["total_delivery_time"] = (
    temporal_analysis["order_delivered_customer_date"]
    - temporal_analysis["order_purchase_timestamp"]
)

In [236]:
temporal_analysis["approval_days"] = (
    temporal_analysis["approval_time"].dt.total_seconds()
    / 86400
)

temporal_analysis["carrier_days"] = (
    temporal_analysis["carrier_time"].dt.total_seconds()
    / 86400
)

temporal_analysis["delivery_days"] = (
    temporal_analysis["delivery_time"].dt.total_seconds()
    / 86400
)

temporal_analysis["total_delivery_days"] = (
    temporal_analysis["total_delivery_time"].dt.total_seconds()
    / 86400
)

In [237]:
temporal_analysis[
    [
        "approval_days",
        "carrier_days",
        "delivery_days",
        "total_delivery_days"
    ]
].describe()

,approval_days,carrier_days,delivery_days,total_delivery_days
count,96455.000000,96455.000000,96455.000000,96455.000000
mean,0.428238,2.799049,9.330407,12.557693
std,0.855673,3.537173,8.760504,9.545982
min,0.000000,-171.219005,-16.096169,0.533414
25%,0.008970,0.874207,4.099670,6.765498
50%,0.014306,1.815810,7.099699,10.216227
75%,0.604890,3.574931,12.029115,15.720145
max,30.893484,125.762569,205.190972,209.628611


In [226]:
temporal_analysis["delivery_late"].value_counts()

delivery_late
False    88630
True      7825
Name: count, dtype: int64

In [238]:
temporal_analysis.groupby("delivery_late")[
    [
        "approval_days",
        "carrier_days",
        "delivery_days",
        "total_delivery_days"
    ]
].mean()

,approval_days,carrier_days,delivery_days,total_delivery_days
delivery_late,,,,
False,0.420741,2.576424,7.886711,10.883875
True,0.513149,5.320612,25.682454,31.516215


In [228]:
temporal_analysis.groupby("delivery_late")[
    [
        "approval_days",
        "carrier_days",
        "delivery_days",
        "total_delivery_days"
    ]
].mean()

,approval_days,carrier_days,delivery_days,total_delivery_days
delivery_late,,,,
False,0.420741,2.576424,7.886711,10.883875
True,0.513149,5.320612,25.682454,31.516215


"Os pedidos classificados como atrasados apresentam tempos significativamente maiores nas etapas de processamento logístico, especialmente entre o envio à transportadora e a entrega ao cliente."

In [239]:
missing_from_temporal = delivered_orders[
    ~delivered_orders["order_id"].isin(temporal_analysis["order_id"])
]

len(missing_from_temporal)

23

In [245]:
missing_from_temporal[
    [
        "order_id",
        "order_status",
        "order_purchase_timestamp",
        "order_approved_at",
        "order_delivered_carrier_date",
        "order_delivered_customer_date",
        "order_estimated_delivery_date"
    ]
]

,order_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date
3002,2d1e2d5bf4dc7227b3bfebb81328c15f,delivered,2017-11-28 17:44:07,2017-11-28 17:56:40,2017-11-30 18:12:23,NaT,2017-12-18
5323,e04abd8149ef81b95221e88f6ed9ab6a,delivered,2017-02-18 14:40:00,NaT,2017-02-23 12:04:47,2017-03-01 13:25:33,2017-03-17
16567,8a9adc69528e1001fc68dd0aaebbb54a,delivered,2017-02-18 12:45:31,NaT,2017-02-23 09:01:52,2017-03-02 10:05:06,2017-03-21
19031,7013bcfc1c97fe719a7b5e05e61c12db,delivered,2017-02-18 13:29:47,NaT,2017-02-22 16:25:25,2017-03-01 08:07:38,2017-03-17
20618,f5dd62b788049ad9fc0526e3ad11a097,delivered,2018-06-20 06:58:43,2018-06-20 07:19:05,2018-06-25 08:05:00,NaT,2018-07-16
22663,5cf925b116421afa85ee25e99b4c34fb,delivered,2017-02-18 16:48:35,NaT,2017-02-22 11:23:10,2017-03-09 07:28:47,2017-03-31
23156,12a95a3c06dbaec84bcfb0e2da5d228a,delivered,2017-02-17 13:05:55,NaT,2017-02-22 11:23:11,2017-03-02 11:09:19,2017-03-20
26800,c1d4211b3dae76144deccd6c74144a88,delivered,2017-01-19 12:48:08,NaT,2017-01-25 14:56:50,2017-01-30 18:16:01,2017-03-01
38290,d69e5d356402adc8cf17e08b5033acfb,delivered,2017-02-19 01:28:47,NaT,2017-02-23 03:11:48,2017-03-02 03:41:58,2017-03-27
39334,d77031d6a3c8a52f019764e68f211c69,delivered,2017-02-18 11:04:19,NaT,2017-02-23 07:23:36,2017-03-02 16:15:23,2017-03-22


In [241]:
negative_carrier = temporal_analysis[
    temporal_analysis["carrier_days"] < 0
]

len(negative_carrier)

1350

In [242]:
negative_carrier[
    [
        "order_id",
        "order_purchase_timestamp",
        "order_approved_at",
        "order_delivered_carrier_date",
        "carrier_days"
    ]
].sort_values("carrier_days")

,order_id,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,carrier_days
25883,7c48bb55e8e4f7e56d412e9653db37bc,2018-07-16 18:40:53,2018-07-16 18:50:22,2018-01-26 13:35:00,-171.219005
14562,1fab4ac9d85079b3da72a11475ae1685,2017-09-01 19:04:22,2017-09-13 22:06:11,2017-09-04 13:10:23,-9.372083
46163,0184d4ddb259e1a4cfc2871888cf97b8,2017-09-01 20:04:28,2017-09-13 22:17:15,2017-09-04 14:05:50,-9.341262
98710,1378f9601350615613cc8832d6789c5d,2017-09-01 20:28:02,2017-09-13 22:03:51,2017-09-04 18:07:55,-9.163843
41592,8554cb37f7158cb0b082a841d24a4589,2017-09-01 18:40:44,2017-09-13 21:58:04,2017-09-04 19:12:19,-9.115104
...,...,...,...,...,...
38841,9256fda5584ec43f6212f63bdf17b2ec,2018-06-27 11:34:34,2018-06-27 11:57:21,2018-06-27 11:56:00,-0.000937
28238,6b3c0661b32905af6722a962d2fac779,2018-07-27 13:30:05,2018-07-27 13:45:13,2018-07-27 13:44:00,-0.000845
38741,b1790ff077b7630f790bba65c012c64b,2018-06-20 15:04:55,2018-06-20 15:40:12,2018-06-20 15:39:00,-0.000833
72849,8d3cd96a636e59a3a47b52cedd6e9bbd,2018-06-20 14:48:16,2018-06-20 15:19:11,2018-06-20 15:18:00,-0.000822


In [243]:
negative_delivery = temporal_analysis[
    temporal_analysis["delivery_days"] < 0
]

len(negative_delivery)

23

In [244]:
negative_delivery[
    [
        "order_id",
        "order_purchase_timestamp",
        "order_approved_at",
        "order_delivered_carrier_date",
        "order_delivered_customer_date",
        "delivery_days"
    ]
].sort_values("delivery_days")

,order_id,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,delivery_days
34939,c1e2bf2b7dd3309f2f5356c6b63968fa,2017-02-10 10:19:10,2017-02-10 10:30:13,2017-03-02 17:34:26,2017-02-14 15:15:57,-16.096169
27470,fa3e37584f4fdb1ded0e0de700dfcb4e,2017-07-30 19:32:23,2017-07-30 19:45:09,2017-08-09 18:18:43,2017-08-01 21:13:01,-7.878958
45302,29941903985f944b0ffc49c479c1547d,2017-05-29 16:16:50,2017-05-29 16:25:16,2017-06-09 15:07:29,2017-06-02 11:09:16,-7.165428
14474,dceb62e8fa94b46006c9554fed743df0,2017-07-20 20:58:05,2017-07-22 11:45:11,2017-08-01 18:23:30,2017-07-26 18:09:10,-6.009954
49933,76458889992169d3135b264dc13aec67,2016-10-07 10:05:16,2016-10-07 11:24:43,2016-10-26 11:43:06,2016-10-20 18:03:17,-5.735984
71227,19feb5627c41ea1b36a8e50a469b3644,2016-10-07 17:09:56,2016-10-07 17:32:09,2016-10-26 11:42:05,2016-10-20 19:07:54,-5.690405
74967,d5558a097766363b8e76b38c43332e8a,2017-02-04 19:01:33,2017-02-04 19:31:12,2017-02-15 08:55:26,2017-02-10 07:58:32,-5.039514
41636,b866af202be0692766081310cd4085e1,2017-01-27 14:59:17,2017-01-27 15:30:46,2017-02-20 02:32:08,2017-02-15 03:53:46,-4.943310
6437,a1abeb653a4d4cd1e142ccb8c82cd069,2017-07-20 11:20:52,2017-07-21 06:43:14,2017-07-28 16:57:58,2017-07-25 19:32:56,-2.892384
78556,ea1dcb4757a844d2642547797bd5feb0,2017-07-18 13:38:29,2017-07-18 13:50:23,2017-07-27 19:21:31,2017-07-25 19:43:10,-1.984965


A inconsistência existe, mas sua incidência é extremamente baixa. Portanto, ela representa principalmente um problema de completude/registro dos dados, e não uma evidência de falha generalizada na operação logística

A análise temporal demonstra que a grande maioria dos pedidos possui informações suficientes para reconstrução do ciclo logístico, porém foram identificadas inconsistências pontuais de completude e inconsistências cronológicas, especialmente no intervalo entre aprovação e encaminhamento à transportadora. Paralelamente, foram identificados atrasos reais de entrega, cuja distribuição apresenta forte assimetria e uma cauda de casos com atrasos muito elevados.

In [246]:
print(
    f"Carrier negativo: {len(negative_carrier)} "
    f"({len(negative_carrier) / len(temporal_analysis) * 100:.2f}%)"
)

print(
    f"Delivery negativo: {len(negative_delivery)} "
    f"({len(negative_delivery) / len(temporal_analysis) * 100:.2f}%)"
)

print(
    f"Registros incompletos: {len(missing_from_temporal)} "
    f"({len(missing_from_temporal) / len(delivered_orders) * 100:.2f}%)"
)

Carrier negativo: 1350 (1.40%)
Delivery negativo: 23 (0.02%)
Registros incompletos: 23 (0.02%)


A qualidade temporal do dataset é, de maneira geral, adequada para análise, pois apenas 0,02% dos pedidos delivered apresentam dados insuficientes para reconstrução completa do fluxo. Entretanto, foram identificadas inconsistências cronológicas em 1,40% dos registros no intervalo entre aprovação e encaminhamento à transportadora, incluindo casos extremos que exigem cautela na interpretação dessas métricas. As inconsistências entre transportadora e entrega ao cliente representam apenas 0,02% dos registros.

Além das inconsistências de dados, a auditoria identificou 7.825 pedidos entregues após a data estimada, mostrando que o atraso de entrega é um fenômeno operacional que merece ser analisado estatisticamente na próxima etapa.

In [247]:
extreme_delays = temporal_analysis[
    temporal_analysis["delay_days"] > 30
]

len(extreme_delays)

360

In [248]:
extreme_delays[
    [
        "order_id",
        "order_purchase_timestamp",
        "order_estimated_delivery_date",
        "order_delivered_customer_date",
        "delay_days"
    ]
].sort_values(
    "delay_days",
    ascending=False
).head(20)

,order_id,order_purchase_timestamp,order_estimated_delivery_date,order_delivered_customer_date,delay_days
55619,1b3190b2dfa9d789e1f14c05b647a14a,2018-02-23 14:57:35,2018-03-15,2018-09-19 23:24:07,188.975081
19590,ca07593549f1816d26a572e06dc1eab6,2017-02-21 23:31:27,2017-03-22,2017-09-19 14:36:39,181.608785
11399,47b40429ed8cce3aee9199792275433f,2018-01-03 09:44:01,2018-01-19,2018-07-13 20:51:31,175.869109
81401,2fe324febf907e3ea3f2aa9650869fa5,2017-03-13 20:17:10,2017-04-05,2017-09-19 17:00:07,167.708414
89130,285ab9426d6982034523a855f55a885e,2017-03-08 22:47:40,2017-04-06,2017-09-19 14:00:04,166.583380
61610,440d0d17af552815d15a9e41abe49359,2017-03-07 23:59:51,2017-04-07,2017-09-19 15:12:50,165.633912
68769,c27815f7e3dd0b926b58552628481575,2017-03-15 23:23:17,2017-04-10,2017-09-19 17:14:25,162.718345
40847,d24e8541128cea179a11a65176e0a96f,2017-06-12 13:14:11,2017-06-26,2017-12-04 18:36:29,161.775336
38509,0f4519c5f1c541ddec9f21b3bddd533a,2017-03-09 13:26:57,2017-04-11,2017-09-19 14:38:21,161.609965
54480,2d7561026d542c8dbd8f0daeadf67a43,2017-03-15 11:24:27,2017-04-13,2017-09-19 14:38:18,159.609931


In [256]:
temporal_analysis["purchase_year"] = (
    temporal_analysis["order_purchase_timestamp"].dt.year
)

temporal_analysis["purchase_month"] = (
    temporal_analysis["order_purchase_timestamp"].dt.month
)

In [257]:
temporal_analysis.groupby(
    "purchase_year"
)["delivery_late"].mean().sort_index()

purchase_year
2016    0.014981
2017    0.066274
2018    0.093677
Name: delivery_late, dtype: float64

In [251]:
temporal_analysis.groupby(
    "purchase_month"
)["delivery_late"].mean().sort_index()

purchase_month
1     0.062300
2     0.134334
3     0.171536
4     0.059554
5     0.066446
6     0.022099
7     0.040786
8     0.075778
9     0.052048
10    0.050601
11    0.143112
12    0.083787
Name: delivery_late, dtype: float64

In [252]:
stage_comparison = temporal_analysis.groupby(
    "delivery_late"
)[
    [
        "approval_days",
        "carrier_days",
        "delivery_days"
    ]
].mean()

stage_comparison

,approval_days,carrier_days,delivery_days
delivery_late,,,
False,0.420741,2.576424,7.886711
True,0.513149,5.320612,25.682454


In [253]:
stage_comparison.loc[True] - stage_comparison.loc[False]

approval_days     0.092408
carrier_days      2.744188
delivery_days    17.795743
dtype: float64

In [254]:
difference = (
    stage_comparison.loc[True]
    - stage_comparison.loc[False]
)

difference.sort_values(ascending=False)

delivery_days    17.795743
carrier_days      2.744188
approval_days     0.092408
dtype: float64

In [255]:
temporal_analysis[
    temporal_analysis["delay_days"] > 30
][
    [
        "order_id",
        "order_purchase_timestamp",
        "order_estimated_delivery_date",
        "order_delivered_customer_date",
        "approval_days",
        "carrier_days",
        "delivery_days",
        "total_delivery_days",
        "delay_days"
    ]
].sort_values(
    "delay_days",
    ascending=False
).head(20)

,order_id,order_purchase_timestamp,order_estimated_delivery_date,order_delivered_customer_date,approval_days,carrier_days,delivery_days,total_delivery_days,delay_days
55619,1b3190b2dfa9d789e1f14c05b647a14a,2018-02-23 14:57:35,2018-03-15,2018-09-19 23:24:07,0.012951,3.147836,205.190972,208.351759,188.975081
19590,ca07593549f1816d26a572e06dc1eab6,2017-02-21 23:31:27,2017-03-22,2017-09-19 14:36:39,1.127639,13.467025,195.033947,209.628611,181.608785
11399,47b40429ed8cce3aee9199792275433f,2018-01-03 09:44:01,2018-01-19,2018-07-13 20:51:31,0.032801,33.636956,157.793785,191.463542,175.869109
81401,2fe324febf907e3ea3f2aa9650869fa5,2017-03-13 20:17:10,2017-04-05,2017-09-19 17:00:07,0.000000,3.462998,186.400162,189.863160,167.708414
89130,285ab9426d6982034523a855f55a885e,2017-03-08 22:47:40,2017-04-06,2017-09-19 14:00:04,0.000000,0.420602,194.213009,194.633611,166.583380
61610,440d0d17af552815d15a9e41abe49359,2017-03-07 23:59:51,2017-04-07,2017-09-19 15:12:50,1.049792,6.492072,188.092153,195.634016,165.633912
68769,c27815f7e3dd0b926b58552628481575,2017-03-15 23:23:17,2017-04-10,2017-09-19 17:14:25,0.000000,5.643843,182.100000,187.743843,162.718345
40847,d24e8541128cea179a11a65176e0a96f,2017-06-12 13:14:11,2017-06-26,2017-12-04 18:36:29,0.007639,4.055162,171.161019,175.223819,161.775336
38509,0f4519c5f1c541ddec9f21b3bddd533a,2017-03-09 13:26:57,2017-04-11,2017-09-19 14:38:21,0.000000,12.667558,181.382025,194.049583,161.609965
54480,2d7561026d542c8dbd8f0daeadf67a43,2017-03-15 11:24:27,2017-04-13,2017-09-19 14:38:18,0.000000,0.920012,187.214606,188.134618,159.609931


In [7]:
import sys

print(sys.executable)

c:\Users\joana\OneDrive\Área de Trabalho\estatistica-modulo-03-medidas-descrititvas\.venv\Scripts\python.exe


In [8]:
import pandas as pd

print(pd.__version__)

3.0.5


In [14]:
%whos

Variable         Type         Data/Info
---------------------------------------
customers        DataFrame    Shape: (99441, 5)
np               module       <module 'numpy' from 'c:\<...>ges\\numpy\\__init__.py'>
order_items      DataFrame    Shape: (112650, 7)
order_payments   DataFrame    Shape: (103886, 5)
orders           DataFrame    Shape: (99441, 8)
pd               module       <module 'pandas' from 'c:<...>es\\pandas\\__init__.py'>
plt              module       <module 'matplotlib.pyplo<...>\\matplotlib\\pyplot.py'>
sns              module       <module 'seaborn' from 'c<...>s\\seaborn\\__init__.py'>
sys              module       <module 'sys' (built-in)>


In [15]:
print(orders.shape)

(99441, 8)


In [16]:
print(order_items.shape)
print(order_payments.shape)
print(customers.shape)

(112650, 7)
(103886, 5)
(99441, 5)


In [17]:
orders.columns

Index(['order_id', 'customer_id', 'order_status', 'order_purchase_timestamp',
       'order_approved_at', 'order_delivered_carrier_date',
       'order_delivered_customer_date', 'order_estimated_delivery_date'],
      dtype='str')

In [18]:
order_items.columns

Index(['order_id', 'order_item_id', 'product_id', 'seller_id',
       'shipping_limit_date', 'price', 'freight_value'],
      dtype='str')